File: Exams/Resoluções de Exames by Carlos Martins/time-process-thread-synch/process-thread/Threads.java
/***
 *
 * ISEL, LEIC, Concurrent Programming
 *
 * Test thread creation in Java
 *
 * Carlos Martins, October 2016
 *
 ***/

import java.util.concurrent.*;
import java.io.IOException;


public class Threads {

	private static final int MAX_THREADS = 20000;

	// The event where all threads are blocked.

	private static CountDownLatch waitEvent = new CountDownLatch(1); 

	public static void main(String[] args) throws IOException, InterruptedException {
		int count = 0;
		Thread[] threads = new Thread[MAX_THREADS];

		for (;;) {
			try {
				//Thread t = new WaitForeverThread();
				
				/*
				 * or
				 */
				threads[count] = new Thread(() -> {
					//
					// Wait unconditionally until the waitEvent is set.
					//
			
					do {
						try {
							waitEvent.await();
							break;
						} catch (InterruptedException ie) {}
					} while (true);
				});
				
				//
				// Comment the next line in order to do not create the thread.
				//	
				
				threads[count].start();
			} catch (Throwable ex) {
				System.out.println("\n*** " + ex);
				break;		
			}
			
			if (++count >= MAX_THREADS)
				break;
			else if ((count % 1000) == 0)
				System.out.println(count);
			else if (count % 20 == 0)
				System.out.print('+');
		}
		System.out.println("\n--created threads: " + count);
		System.out.println("--active threads: " + Thread.activeCount());
		System.out.print("---hit <enter> to set the wait event...");
		
		//
		// Wait for <enter> to signal the wait event.
		//
		
		System.in.read();		
		waitEvent.countDown();
		
		//
		// Since all threads were created as non-daemon threads, the
		// process termination only occurs after all threads have been
		// exited.
		//
		
		long start = System.currentTimeMillis();
		
		// synchronize with termination of all threads
		for (int i = 0; i < count; ++i)
			threads[i].join();
		System.out.printf("--wait time: %d ms%n", System.currentTimeMillis() - start);
	}
}

File: Exams/Resoluções de Exames by Carlos Martins/time-process-thread-synch/process-thread/Threads.cs
/***
 *
 * ISEL, LEIC, Concurrent Programming
 *
 * Test thread creation and exit using the .NET framework
 *
 * Carlos Martins, October 2016
 *
 ***/

using System;
using System.Threading;

static public class Threads {

	private static uint MAX_THREADS = 20000;

	private static void Main()
	{
		Thread[] threads = new Thread[MAX_THREADS];

		// The event where all of the created threads will block.
		ManualResetEventSlim waitEvent = new ManualResetEventSlim(false); 
		int count;
		for (count = 0; count < MAX_THREADS; ++count) {
			try {
				(threads[count] = new Thread(() => waitEvent.Wait())).Start();
			} catch (Exception ex) {
				Console.WriteLine(ex.Message);
				break;
			}
			
			if (count != 0 && (count % 1000) == 0)
				Console.WriteLine("{0}", count);
			else if (count != 0 && count % 20 == 0)
				Console.Write('+');
		}
		Console.WriteLine("{0}", count);
		Console.Write("--hit <enter> to set the wait event...");
		Console.ReadLine();
		waitEvent.Set();
		int start = Environment.TickCount;
		// synchronize with termination of all threads	
		for (int i = 0; i < MAX_THREADS; ++i)
			threads[i].Join();
		Console.WriteLine("--wait time: {0} ms", Environment.TickCount - start);
	}
}
